File: problems/1301-1400/1396_design-underground-system/solution.cpp
#include <map>
#include <string>
#include <unordered_map>
#include <utility>

class UndergroundSystem {
  public:
    UndergroundSystem() = default;

    void checkIn(int id, std::string stop, int t) { checkins[id] = {std::move(stop), t}; }

    void checkOut(int id, std::string stop, int t) {
        auto start = checkins.find(id);
        std::string from = std::move(start->second.first);
        long long duration = (long long)t - start->second.second;
        checkins.erase(start);
        auto &bucket = totals[{std::move(from), std::move(stop)}];
        bucket.first += duration;
        bucket.second += 1;
    }

    double getAverageTime(std::string fromStop, std::string toStop) {
        const auto &bucket = totals.at({std::move(fromStop), std::move(toStop)});
        return (double)bucket.first / bucket.second;
    }

  private:
    std::unordered_map<int, std::pair<std::string, int>> checkins;
    std::map<std::pair<std::string, std::string>, std::pair<long long, long long>> totals;
};

File: problems/1301-1400/1396_design-underground-system/solution.go
package main

type tripCheckin struct {
	stop string
	time int
}

type tripKey struct {
	from string
	to   string
}

type tripTotal struct {
	sum   int64
	count int64
}

type UndergroundSystem struct {
	checkins map[int]tripCheckin
	totals   map[tripKey]tripTotal
}

func NewUndergroundSystemTyped() *UndergroundSystem {
	return &UndergroundSystem{}
}

func (design *UndergroundSystem) ensure() {
	if design.checkins == nil {
		design.checkins = make(map[int]tripCheckin)
		design.totals = make(map[tripKey]tripTotal)
	}
}

func (design *UndergroundSystem) checkIn(id int, stop string, t int) {
	design.ensure()
	design.checkins[id] = tripCheckin{stop: stop, time: t}
}

func (design *UndergroundSystem) checkOut(id int, stop string, t int) {
	design.ensure()
	start := design.checkins[id]
	delete(design.checkins, id)
	key := tripKey{from: start.stop, to: stop}
	bucket := design.totals[key]
	bucket.sum += int64(t - start.time)
	bucket.count++
	design.totals[key] = bucket
}

func (design *UndergroundSystem) getAverageTime(fromStop string, toStop string) float64 {
	bucket := design.totals[tripKey{from: fromStop, to: toStop}]
	return float64(bucket.sum) / float64(bucket.count)
}
